File: apps/site/src/lib/board.ts
/**
 * The "Board" — the rich, harness-centric data model the public site renders.
 *
 * POLLMEVALS evaluates STACKS = (model × harness/scaffolding), not bare models.
 * The board is shaped so the central question — *which harness, with which model,
 * is best, and for which task* — is answerable at a glance:
 *
 *   harnesses  → rows of the matrix (ordered L0 → L8 by scaffolding depth)
 *   models     → columns (ordered cheap → expensive)
 *   cells      → one (model × harness) stack, with per-task breakdown
 *
 * Honesty: `illustrative: true` marks a preview dataset built to demonstrate the
 * design before real multi-harness scored runs exist (the executor that runs
 * non-raw harnesses is a separate, cost-gated build). Real runs replace it.
 */

export interface Harness {
  stack_id: string;
  name: string; // display, e.g. "ForgePlan Framework"
  /** Max scaffolding level present (0-8) — drives row order. */
  level: number;
  /** Active L-layers, e.g. ["L1_system_prompt","L2_tools"]. */
  layers: string[];
  /** baseline | agnostic | vendor | framework — for grouping/legend. */
  family: "baseline" | "agnostic" | "vendor" | "framework";
}

export interface ModelRef {
  model_id: string;
  name: string;
  /** cheap | mid | frontier — drives column order (cheap left). */
  tier: "cheap" | "mid" | "frontier";
}

export interface TaskScore {
  score: number | null; // 0-10
  cost_usd: number;
  pass_hat_k: number | null;
}

export interface Cell {
  model_id: string;
  stack_id: string;
  /** Overall mean across tasks (null on unscored). */
  mean_score: number | null;
  mean_cost_usd: number;
  mean_latency_ms: number;
  pass_hat_k: number | null;
  /** quality ÷ cost — the thesis metric (null when unscored or cost 0). */
  quality_per_dollar: number | null;
  per_task: Record<string, TaskScore>;
  on_frontier: boolean;
}

export interface Board {
  run_hash: string;
  run_type: string;
  methodology_version: string;
  region: string;
  scored: boolean;
  illustrative: boolean;
  solved_threshold: number;
  harnesses: Harness[];
  models: ModelRef[];
  tasks: string[];
  cells: Cell[];
}

// ---------------------------------------------------------------------------
// Lookups + derivations
// ---------------------------------------------------------------------------

const key = (modelId: string, stackId: string) => `${modelId}::${stackId}`;

export function cellMap(board: Board): Map<string, Cell> {
  return new Map(board.cells.map((c) => [key(c.model_id, c.stack_id), c]));
}

/** A matrix in render order: rows = harnesses (deepest scaffold first), cols = models (cheap first). */
export interface MatrixView {
  rows: Harness[]; // ordered: highest level first (so scaffold climbs downward visually = top)
  cols: ModelRef[]; // ordered: cheapest first
  cellAt: (h: Harness, m: ModelRef) => Cell | undefined;
}

export function buildMatrix(board: Board): MatrixView {
  const rows = [...board.harnesses].sort((a, b) => b.level - a.level);
  const tierRank = { cheap: 0, mid: 1, frontier: 2 } as const;
  const cols = [...board.models].sort(
    (a, b) => tierRank[a.tier] - tierRank[b.tier]
  );
  const map = cellMap(board);
  return {
    rows,
    cols,
    cellAt: (h, m) => map.get(key(m.model_id, h.stack_id)),
  };
}

/** What a matrix cell is colored by. */
export type Metric = "quality_per_dollar" | "mean_score" | "mean_cost_usd";

export function metricValue(cell: Cell, metric: Metric): number | null {
  if (metric === "mean_cost_usd") return cell.mean_cost_usd;
  return cell[metric];
}

/** The single best stack under a metric (max, or min for cost). */
export function bestCell(board: Board, metric: Metric): Cell | null {
  const valued = board.cells
    .map((c) => ({ c, v: metricValue(c, metric) }))
    .filter((x): x is { c: Cell; v: number } => x.v !== null);
  if (valued.length === 0) return null;
  const pick =
    metric === "mean_cost_usd"
      ? valued.reduce((a, b) => (b.v < a.v ? b : a))
      : valued.reduce((a, b) => (b.v > a.v ? b : a));
  return pick.c;
}

/** Min/max of a metric across scored cells — for the heatmap color scale. */
export function metricRange(
  board: Board,
  metric: Metric
): { min: number; max: number } {
  const vals = board.cells
    .map((c) => metricValue(c, metric))
    .filter((v): v is number => v !== null);
  if (vals.length === 0) return { min: 0, max: 1 };
  return { min: Math.min(...vals), max: Math.max(...vals) };
}

/**
 * Pareto frontier over cells: maximize quality, minimize cost. Unscored cells
 * (no mean_score) are excluded — there's no quality axis to be on the frontier of.
 * Mutates nothing; returns the set of frontier keys.
 */
export function frontierKeys(board: Board): Set<string> {
  const scored = board.cells.filter((c) => c.mean_score !== null);
  const out = new Set<string>();
  for (const c of scored) {
    const dominated = scored.some(
      (q) =>
        q !== c &&
        q.mean_cost_usd <= c.mean_cost_usd &&
        (q.mean_score ?? 0) >= (c.mean_score ?? 0) &&
        (q.mean_cost_usd < c.mean_cost_usd ||
          (q.mean_score ?? 0) > (c.mean_score ?? 0))
    );
    if (!dominated) out.add(key(c.model_id, c.stack_id));
  }
  return out;
}

/** Per task: the highest-scoring stack (cell) — different tasks, different winners. */
export interface TaskWinner {
  task: string;
  cell: Cell;
  score: number;
}

export function perTaskWinners(board: Board): TaskWinner[] {
  return board.tasks
    .map((task) => {
      let best: { cell: Cell; score: number } | null = null;
      for (const c of board.cells) {
        const s = c.per_task[task]?.score;
        if (
          s !== null &&
          s !== undefined &&
          (best === null || s > best.score)
        ) {
          best = { cell: c, score: s };
        }
      }
      return best ? { task, cell: best.cell, score: best.score } : null;
    })
    .filter((w): w is TaskWinner => w !== null);
}

/**
 * Scaffolding-lift series: for each model, its cells ordered by harness level —
 * the climb from L0 (bare) up the ladder. Powers the per-model ablation view.
 */
export interface LiftPoint {
  harness: Harness;
  cell: Cell;
}
export interface LiftSeries {
  model: ModelRef;
  points: LiftPoint[]; // ordered by harness level ascending (L0 first)
}

export function scaffoldingLift(board: Board): LiftSeries[] {
  const map = cellMap(board);
  const byLevel = [...board.harnesses].sort((a, b) => a.level - b.level);
  return board.models.map((model) => ({
    model,
    points: byLevel
      .map((harness) => {
        const cell = map.get(key(model.model_id, harness.stack_id));
        return cell ? { harness, cell } : null;
      })
      .filter((p): p is LiftPoint => p !== null),
  }));
}

File: apps/eval-core-py/src/leaderboard/board.py
"""Emit the rich harness x model Board (apps/site/src/lib/board.ts shape) from runs.

RFC-006 Phase 4c. Turns scored ``EvalRow``s into the JSON the public site
renders, replacing the illustrative preview (``board.illustrative.json``) with
real ``(model x harness x task)`` numbers.

Mirrors the TypeScript ``Board`` contract field-for-field. Harness metadata
(level, family, layers) is derived from ``stacks/<id>/stack.yaml``; model
metadata (name, tier) from a small registry; cell score/cost/latency from the
EvalRows (judge-aggregate median when a deterministic score isn't set).
"""

from __future__ import annotations

import statistics
from collections import defaultdict
from pathlib import Path

import yaml
from pydantic import BaseModel, ConfigDict

from src.contracts import EvalRow, EvalStatus

# ---------------------------------------------------------------------------
# Harness / model metadata (the Board needs display fields the run doesn't carry)
# ---------------------------------------------------------------------------

# Scaffolding family per stack (matches board.ts Harness.family). Default
# "agnostic" for unknown CLI harnesses.
_STACK_FAMILY: dict[str, str] = {
    "raw-llm": "baseline",
    "aider": "agnostic",
    "codex": "agnostic",
    "opencode": "agnostic",
    "goose": "agnostic",
    "openhands": "agnostic",
    "hermes": "agnostic",
    "cline": "vendor",
    "claude-code-basic": "vendor",
    "pi": "vendor",
    "forgeplan-framework": "framework",
}

# Ordered L-keys → numeric level. The harness level is the highest true layer.
_LAYER_ORDER: list[str] = [
    "L0_bare_llm",
    "L1_system_prompt",
    "L2_tools",
    "L3_skills",
    "L4_file_memory",
    "L5_vector_memory",
    "L6_subagents",
    "L7_validator",
    "L8_framework",
]

# Model display name + cost tier, keyed by the proxy alias / model_id segment.
_MODEL_REGISTRY: dict[str, tuple[str, str]] = {
    "qwen-3-14b": ("Qwen 3 14B", "cheap"),
    "qwen-3-32b": ("Qwen 3 32B", "cheap"),
    "llama-3-3-70b": ("Llama 3.3 70B", "cheap"),
    "gemini-3-flash": ("Gemini 3 Flash", "mid"),
    "gpt-5-mini": ("GPT-5 mini", "mid"),
    "deepseek-v3-5": ("DeepSeek V3.5", "mid"),
    "claude-sonnet-4-6": ("Claude Sonnet 4.6", "frontier"),
    "gpt-5": ("GPT-5", "frontier"),
    "claude-opus-4-7": ("Claude Opus 4.7", "frontier"),
    "gemini-2-5-pro": ("Gemini 2.5 Pro", "frontier"),
}


def _alias(model_id: str) -> str:
    """provider-route → short alias (matches default_model_alias)."""
    return model_id.rsplit("/", 1)[-1]


# ---------------------------------------------------------------------------
# Board contract (mirrors apps/site/src/lib/board.ts)
# ---------------------------------------------------------------------------


class Harness(BaseModel):
    model_config = ConfigDict(extra="forbid")
    stack_id: str
    name: str
    level: int
    layers: list[str]
    family: str


class ModelRef(BaseModel):
    model_config = ConfigDict(extra="forbid")
    model_id: str
    name: str
    tier: str


class TaskScore(BaseModel):
    model_config = ConfigDict(extra="forbid")
    score: float | None
    cost_usd: float
    pass_hat_k: float | None = None


class Cell(BaseModel):
    model_config = ConfigDict(extra="forbid")
    model_id: str
    stack_id: str
    mean_score: float | None
    mean_cost_usd: float
    mean_latency_ms: int
    pass_hat_k: float | None = None
    quality_per_dollar: float | None
    per_task: dict[str, TaskScore]
    on_frontier: bool = False


class Board(BaseModel):
    model_config = ConfigDict(extra="forbid")
    run_hash: str
    run_type: str
    methodology_version: str
    region: str
    scored: bool
    illustrative: bool
    solved_threshold: float
    harnesses: list[Harness]
    models: list[ModelRef]
    tasks: list[str]
    cells: list[Cell]


# ---------------------------------------------------------------------------
# Derivations
# ---------------------------------------------------------------------------


def _harness_from_yaml(stack_id: str, stacks_root: Path) -> Harness:
    """Build Harness display metadata from stacks/<id>/stack.yaml."""
    data = yaml.safe_load((stacks_root / stack_id / "stack.yaml").read_text(encoding="utf-8"))
    layers_map = data.get("layers", {}) if isinstance(data, dict) else {}
    true_layers = [k for k in _LAYER_ORDER if isinstance(layers_map, dict) and layers_map.get(k)]
    level = max((_LAYER_ORDER.index(k) for k in true_layers), default=0)
    return Harness(
        stack_id=stack_id,
        name=str(data.get("name", stack_id)) if isinstance(data, dict) else stack_id,
        level=level,
        layers=true_layers,
        family=_STACK_FAMILY.get(stack_id, "agnostic"),
    )


def _model_ref(model_id: str) -> ModelRef:
    name, tier = _MODEL_REGISTRY.get(_alias(model_id), (_alias(model_id), "mid"))
    return ModelRef(model_id=model_id, name=name, tier=tier)


def _row_score(row: EvalRow) -> float | None:
    """Single 0-10 score for an EvalRow: final_score, else judge-aggregate mean."""
    if row.final_score is not None:
        return round(float(row.final_score), 2)
    agg = row.judge_aggregate
    if agg is not None and agg.median_per_criterion:
        return round(statistics.mean(agg.median_per_criterion.values()), 2)
    return None


# ---------------------------------------------------------------------------
# Emitter
# ---------------------------------------------------------------------------


def build_board(
    eval_rows: list[EvalRow],
    *,
    stacks_root: Path,
    run_hash: str,
    run_type: str = "weekly",
    methodology_version: str = "v0.1.0",
    region: str = "eu-central",
    solved_threshold: float = 6.0,
    illustrative: bool = False,
) -> Board:
    """Aggregate scored EvalRows into the public Board.

    One cell per (model_id, stack_id). per_task holds each task's score/cost;
    mean_* average across the cell's tasks. quality_per_dollar = mean_score /
    mean_cost. ``on_frontier`` is left False (the site computes it client-side).
    FAILED rows contribute cost (transparency) but a null score.
    """
    # group rows by (model_id, stack_id) then task_id
    by_cell: dict[tuple[str, str], dict[str, list[EvalRow]]] = defaultdict(
        lambda: defaultdict(list)
    )
    for r in eval_rows:
        by_cell[(r.model_id, r.stack_id)][r.task_id].append(r)

    model_ids = sorted({m for m, _ in by_cell})
    stack_ids = sorted({s for _, s in by_cell})
    task_ids = sorted({r.task_id for r in eval_rows})

    cells: list[Cell] = []
    any_scored = False
    for (model_id, stack_id), tasks in sorted(by_cell.items()):
        per_task: dict[str, TaskScore] = {}
        for task_id, rows in tasks.items():
            scores = [s for r in rows if (s := _row_score(r)) is not None]
            costs = [float(r.stats.cost_usd) for r in rows]
            per_task[task_id] = TaskScore(
                score=round(statistics.median(scores), 2) if scores else None,
                cost_usd=round(statistics.mean(costs), 6) if costs else 0.0,
            )
        cell_scores = [t.score for t in per_task.values() if t.score is not None]
        cell_costs = [t.cost_usd for t in per_task.values()]
        all_rows = [r for rows in tasks.values() for r in rows]
        latencies = [r.stats.wall_clock_ms for r in all_rows]
        mean_score = round(statistics.mean(cell_scores), 2) if cell_scores else None
        mean_cost = round(statistics.mean(cell_costs), 6) if cell_costs else 0.0
        if mean_score is not None:
            any_scored = True
        cells.append(
            Cell(
                model_id=model_id,
                stack_id=stack_id,
                mean_score=mean_score,
                mean_cost_usd=mean_cost,
                mean_latency_ms=int(statistics.mean(latencies)) if latencies else 0,
                quality_per_dollar=(
                    round(mean_score / mean_cost, 2)
                    if mean_score is not None and mean_cost > 0
                    else None
                ),
                per_task=per_task,
            )
        )

    return Board(
        run_hash=run_hash,
        run_type=run_type,
        methodology_version=methodology_version,
        region=region,
        scored=any_scored,
        illustrative=illustrative,
        solved_threshold=solved_threshold,
        harnesses=[_harness_from_yaml(s, stacks_root) for s in stack_ids],
        models=[_model_ref(m) for m in model_ids],
        tasks=task_ids,
        cells=cells,
    )


def _statuses_present(eval_rows: list[EvalRow]) -> set[EvalStatus]:
    """Helper for callers/tests: which EvalStatus values are in the set."""
    return {r.status for r in eval_rows}
